File: src/handlers/actions.py
"""Interactive action handlers for WeeklyRoulette bot."""

from slack_bolt import Ack, App
from slack_sdk import WebClient

from database.models import DAY_DISPLAY_NAMES, ChannelConfig
from services.roulette import RouletteService
from services.scheduler import SchedulerService


def register_action_handlers(
    app: App, roulette_service: RouletteService, scheduler_service: SchedulerService
) -> None:
    """Register all interactive action handlers."""

    @app.view("config_modal")
    def handle_config_modal_submission(ack: Ack, body: dict, client: WebClient) -> None:
        """Handle configuration modal submission."""
        ack()

        try:
            # Extract form data
            view = body["view"]
            channel_id = view["private_metadata"]
            values = view["state"]["values"]

            # Get selected values
            day_block = next(iter(values.values()))
            day = day_block["day_select"]["selected_option"]["value"]

            time_block = list(values.values())[1]
            time = time_block["time_select"]["selected_option"]["value"]

            enabled_block = list(values.values())[2]
            enabled = (
                enabled_block["enabled_select"]["selected_option"]["value"] == "true"
            )

            # Create and save configuration
            config = ChannelConfig(
                channel_id=channel_id, day=day, time=time, enabled=enabled
            )

            roulette_service.db.save_channel_config(config)

            # Update scheduler
            scheduler_service.force_update_schedules()

            # Send confirmation message
            day_name = DAY_DISPLAY_NAMES[day]
            status_emoji = "✅" if enabled else "⏸️"

            confirmation_message = (
                f"{status_emoji} **Configuration saved!**\n\n"
                f"📅 Day: {day_name}\n"
                f"🕒 Time: {time}\n"
            )

            if enabled:
                confirmation_message += (
                    f"🎯 First automatic selection will happen on "
                    f"next {day_name.lower()} at {time}.\n\n"
                )
            else:
                confirmation_message += (
                    "⚠️ Automatic selection is disabled. "
                    "Use `/weeklyroulette config` again to enable it."
                )

            client.chat_postMessage(channel=channel_id, text=confirmation_message)

            print(
                f"✅ Configuration saved for channel {channel_id}: "
                f"{day} at {time} (enabled: {enabled})"
            )

        except Exception as e:
            print(f"❌ Error handling config modal submission: {e}")

            # Send error message
            try:
                client.chat_postMessage(
                    channel=channel_id,
                    text="❌ An error occurred while saving configuration. Please try again.",
                )
            except Exception:
                pass

    @app.action("day_select")
    def handle_day_select(ack: Ack) -> None:
        """Handle day selection in config modal."""
        ack()

    @app.action("time_select")
    def handle_time_select(ack: Ack) -> None:
        """Handle time selection in config modal."""
        ack()

    @app.action("enabled_select")
    def handle_enabled_select(ack: Ack) -> None:
        """Handle enabled/disabled selection in config modal."""
        ack()

    print("✅ Action handlers registered")

File: internal/handlers/actions.go
package handlers

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/slack-go/slack"
	"github.com/typeWolffo/weeklyroulette-bot/internal/database"
	"github.com/typeWolffo/weeklyroulette-bot/internal/services"
)

type ActionHandler struct {
	roulette  *services.RouletteService
	scheduler *services.SchedulerService
}

func NewActionHandler(roulette *services.RouletteService, scheduler *services.SchedulerService) *ActionHandler {
	return &ActionHandler{
		roulette:  roulette,
		scheduler: scheduler,
	}
}

func (h *ActionHandler) HandleConfigModalSubmission(view slack.View, client *slack.Client) error {
	channelID := view.PrivateMetadata
	values := view.State.Values

	var day, timeStr string
	var enabled bool

	for _, blockValues := range values {
		if selectVal, ok := blockValues["day_select"]; ok {
			day = selectVal.SelectedOption.Value
		}
		if selectVal, ok := blockValues["time_select"]; ok {
			timeStr = selectVal.SelectedOption.Value
		}
		if selectVal, ok := blockValues["enabled_select"]; ok {
			enabled = selectVal.SelectedOption.Value == "true"
		}
	}

	if day == "" || timeStr == "" {
		return fmt.Errorf("missing required fields")
	}

	ctx := context.Background()
	config := &database.ChannelConfig{
		ChannelID: channelID,
		Day:       day,
		Time:      timeStr,
		Enabled:   enabled,
	}

	if err := h.roulette.DB().SaveChannelConfig(ctx, config); err != nil {
		slog.Error("failed to save channel config", "error", err)
		return fmt.Errorf("saving config: %w", err)
	}

	h.scheduler.ForceUpdateSchedules()

	statusText := "enabled"
	if !enabled {
		statusText = "disabled"
	}

	dayName := database.DayDisplayNames[day]
	message := fmt.Sprintf(
		"✅ Configuration saved!\n\n"+
			"📅 Day: %s\n"+
			"🕒 Time: %s (Polish time)\n"+
			"🔄 Status: %s",
		dayName, timeStr, statusText,
	)

	_, _, err := client.PostMessage(channelID,
		slack.MsgOptionText(message, false),
	)
	if err != nil {
		slog.Warn("failed to send confirmation message", "error", err)
	}

	slog.Info("channel config saved",
		"channel", channelID,
		"day", day,
		"time", timeStr,
		"enabled", enabled,
	)

	return nil
}

func (h *ActionHandler) HandleBlockAction(action *slack.BlockAction) {
	slog.Debug("block action received", "action_id", action.ActionID, "value", action.SelectedOption.Value)
}
